### PR TITLE
Resolve nested dependencies using dependents repo hierarchy

### PR DESCRIPTION
**Pull request by hemikak**

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java [hunk 1]
```diff
@@ -283,14 +290,15 @@ public BLangPackage loadEntryPackage(PackageID pkgId, PackageID enclPackageId, b
         return packageNode;
     }
 
-    public BLangPackage loadPackage(PackageID pkgId, PackageID enclPackageId, PackageRepository packageRepo) {
+    public BLangPackage loadPackage(PackageID pkgId, PackageID enclPackageId, RepoHierarchy encPkgRepoHierarchy) {
```
Comment (anupama-pathirage) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java]: This method can be private?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java [hunk 2]
```diff
@@ -283,14 +290,15 @@ public BLangPackage loadEntryPackage(PackageID pkgId, PackageID enclPackageId, b
...
         // TODO Remove this method()
         BLangPackage bLangPackage = packageCache.get(pkgId);
         if (bLangPackage != null) {
             return bLangPackage;
         }
 
-        BLangPackage packageNode = loadPackageFromEntity(pkgId, loadPackageEntity(pkgId, enclPackageId));
+        BLangPackage packageNode = loadPackageFromEntity(pkgId, loadPackageEntity(pkgId, enclPackageId,
+                                                                                                encPkgRepoHierarchy));
```
Comment (anupama-pathirage) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java]: Shall we reformat this code?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/packaging/Resolution.java [hunk 3]
```diff
@@ -9,11 +10,11 @@
  * Had to wrap in a class since you can't return multiple items form a java method.
  */
 public class Resolution {
-    public static final Resolution NOT_FOUND = new Resolution(null, null);
+    public static final Resolution NOT_FOUND = new Resolution(null, null, null);
     public final RepoHierarchy resolvedBy;
     public final List<CompilerInput> inputs;
 
-    Resolution(RepoHierarchy resolvedBy, List<CompilerInput> inputs) {
+    Resolution(RepoHierarchy resolvedBy, List<CompilerInput> inputs, Repo resolvedRepository) {
```
Comment (anupama-pathirage) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/packaging/Resolution.java]: Why passing unused variable here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/packaging/Resolution.java [hunk 4]
```diff
@@ -9,11 +10,11 @@
  * Had to wrap in a class since you can't return multiple items form a java method.
  */
 public class Resolution {
-    public static final Resolution NOT_FOUND = new Resolution(null, null);
+    public static final Resolution NOT_FOUND = new Resolution(null, null, null);
```
Comment (anupama-pathirage) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/packaging/Resolution.java]: We don't need to pass last null as it is not used

File: tests/ballerina-tools-integration-test/src/test/resources/repohierarchy/x/main.bal [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import integrationtests/b;
+import ballerina/io;
```
Comment (anupama-pathirage) [inline tests/ballerina-tools-integration-test/src/test/resources/repohierarchy/x/main.bal]: Unused import?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java (hunk 1 above)
Comment (hemikak) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java]: Fixed with https://github.com/ballerina-platform/ballerina-lang/pull/10442/commits/fa74255bc087ed97ca3d7fb2f28539751a69e5f0

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java (hunk 2 above)
Comment (hemikak) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java]: Fixed with https://github.com/ballerina-platform/ballerina-lang/pull/10442/commits/fa74255bc087ed97ca3d7fb2f28539751a69e5f0

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/packaging/Resolution.java (hunk 3 above)
Comment (hemikak) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/packaging/Resolution.java]: Fixed with https://github.com/ballerina-platform/ballerina-lang/pull/10442/commits/fa74255bc087ed97ca3d7fb2f28539751a69e5f0

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/packaging/Resolution.java (hunk 4 above)
Comment (hemikak) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/packaging/Resolution.java]: Fixed with https://github.com/ballerina-platform/ballerina-lang/pull/10442/commits/fa74255bc087ed97ca3d7fb2f28539751a69e5f0

File: tests/ballerina-tools-integration-test/src/test/resources/repohierarchy/x/main.bal (hunk 5 above)
Comment (hemikak) [reply tests/ballerina-tools-integration-test/src/test/resources/repohierarchy/x/main.bal]: Fixed with https://github.com/ballerina-platform/ballerina-lang/pull/10442/commits/fa74255bc087ed97ca3d7fb2f28539751a69e5f0

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java [hunk 6]
```diff
@@ -283,14 +290,14 @@ public BLangPackage loadEntryPackage(PackageID pkgId, PackageID enclPackageId, b
...
+    private BLangPackage loadPackage(PackageID pkgId) {
         // TODO Remove this method()
         BLangPackage bLangPackage = packageCache.get(pkgId);
         if (bLangPackage != null) {
             return bLangPackage;
         }
-
-        BLangPackage packageNode = loadPackageFromEntity(pkgId, loadPackageEntity(pkgId, enclPackageId));
+    
+        BLangPackage packageNode = loadPackageFromEntity(pkgId, loadPackageEntity(pkgId, null, null));
```
Comment (manuranga) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java]: can we introduce a new method instead passing null, can be confusing to read.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/packaging/converters/PathConverter.java [hunk 7]
```diff
@@ -94,7 +96,14 @@ public Path start() {
     }
 
     @Override
-    public Stream<CompilerInput> finalize(Path path, PackageID id) {
+    public Stream<CompilerInput> finalize(Path path, PackageID pkgId) {
+        // Set package version if its empty
+        if (pkgId.version.value.isEmpty() && !pkgId.orgName.equals(Names.BUILTIN_ORG)
+                                                                    && !pkgId.orgName.equals(Names.ANON_ORG)) {
```
[the remark this answers is not in the document]
Comment (hemikak) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/packaging/converters/PathConverter.java]: Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/10442/commits/dffb67936effdf7af12306849aa3c093cefc19d7

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java (hunk 6 above)
Comment (hemikak) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/PackageLoader.java]: Created a separate method, Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/10442/commits/dffb67936effdf7af12306849aa3c093cefc19d7

**Pull request merged**